### PR TITLE
Add decoder test framework

### PR DESCRIPTION
**Pull request by szkarpinski**
## Category:

**New feature** (*non-breaking change which adds functionality*)

## Description:

This PR adds a framework for testing imgcodec's decoders. It allows comparing results of decoding with reference image stored in [numpy format](https://numpy.org/doc/stable/reference/generated/numpy.lib.format.html). The general idea is that reference numpy files with decoded images will be stored in `DALI_extra`.

**Why numpy format?** Files in numpy format are universal, self-contained, easy to create and view, and we already have C++ reader code. Another alternative would be comparing against trusted (?) library such as OpenCV, but this would give us no control over the reference output and could lead to situations where tests stop working because of breaking OpenCV change.

### Future work:
The following things are out of scope of this PR:
- At some point we'd like to generalize it for GPU codecs

### Example:

When writing a test for decoder, one should override two methods, which would instantiate the parser and decoder. The example usage, taken from https://github.com/NVIDIA/DALI/pull/4109 is presented below.

```cpp
auto img_ref_path = dali_extra + "/db/single/reference/tiff/0/cat-111793_640.tiff.npy";
auto img_path = dali_extra + "/db/single/tiff/0/cat-111793_640.tiff";

class LibTiffDecoderTest : public NumpyDecoderTestBase<uint8_t> {
 protected:
  std::shared_ptr<ImageDecoderInstance> CreateDecoder(ThreadPool &tp) override {
    LibTiffDecoder decoder;
    return decoder.Create(CPU_ONLY_DEVICE_ID, tp);
  }
  std::shared_ptr<ImageParser> CreateParser() override {
    return std::make_shared<TiffParser>();
  }
};

TEST_F(LibTiffDecoderTest, Test) {
  auto ref = ReadReferenceFrom(img_ref_path);
  auto src = ImageSource::FromFilename(img_path);
  auto img = Decode(&src);
  AssertEqual(img, ref);
}

TEST_F(LibTiffDecoderTest, TestROI) {
  auto ref = ReadReferenceFrom(img_ref_path);
  auto src = ImageSource::FromFilename(img_path);
  auto info = Parser()->GetInfo(&src);

  ROI roi = {{13, 17, 0}, {info.shape[0] - 55, info.shape[1] - 10, 3}};
  auto img = Decode(&src, {}, roi);
  AssertEqual(img, Crop(ref, roi));
}
```

### Affected modules and functionalities:
- The framework lives in `decoder_test_helper.h` and provides the `NumpyDecoderTestBase` class template

### Key points relevant for the review:
This PR is kind of a blocker for PRs with codecs. I'd be happy to add any missing features or improve the efficiency of the tests, but I'd prefer to do so in follow-up PRs, to letothers write their codecs.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2908

### CODE REVIEW

File: dali/imgcodec/decoders/decoder_test_helper.h [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+  /**
+  * @brief Checks if two tensors are equal.
+  */
+  void AssertEqual(const Tensor<CPUBackend> &img, const Tensor<CPUBackend> &ref) {
+    EXPECT_EQ(img.shape(), ref.shape()) << "Different shapes";
+    auto img_view = view<const OutputType>(img), ref_view = view<const OutputType>(ref);
+    for (int i = 0; i < volume(img.shape()); i++) {
+      // TODO(skarpinski) Pretty-print position on error
+      ASSERT_EQ(img_view.data[i], ref_view.data[i]) << "at " << PrettyPosition(i, img.shape());
+    }
```
Comment (mzient) [inline dali/imgcodec/decoders/decoder_test_helper.h:74]:
```suggestion
    Check(view<const OutputType>(img), view<const OutputType>(ref));
```

from `dali/test/tensor_test_utils.h`

File: dali/imgcodec/decoders/decoder_test_helper.h [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    result << "(";
+    for (size_t i = 0; i < pos.size(); i++) {
+      result << pos[i];
+      if (i < pos.size() - 1)
+        result << ", ";
+      else
+        result << ")";
+    }
+    return result.str();
+  }
```
Comment (mzient) [inline dali/imgcodec/decoders/decoder_test_helper.h:160]: Not necessary (use Check, it already does what's needed).

File: dali/imgcodec/decoders/decoder_test_helper.h [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+      Tensor<CPUBackend> transposed;
+      transposed.Resize(data.shape(), data.type());
+      SampleView<CPUBackend> input(data.raw_mutable_data(), data.shape(), data.type());
+      SampleView<CPUBackend> output(transposed.raw_mutable_data(), transposed.shape(),
+                                    transposed.type());
+      numpy::FromFortranOrder(output, input);
+      return transposed;
+    }
+    return data;
+  }
```
Comment (mzient) [inline dali/imgcodec/decoders/decoder_test_helper.h:213]: Move this to `numpy` header or generic test headers? In any case, this should net be a member function of any class.

File: dali/imgcodec/decoders/decoder_test_helper.h (hunk 1 above)
Comment (szkarpinski) [reply dali/imgcodec/decoders/decoder_test_helper.h:74]: Wow, thanks!

File: dali/imgcodec/decoders/decoder_test_helper.h (hunk 3 above)
Comment (szkarpinski) [reply dali/imgcodec/decoders/decoder_test_helper.h:213]: Yeah, probably `numpy.h` would be a better place for that

File: dali/imgcodec/decoders/decoder_test_helper.h (hunk 1 above)
Comment (szkarpinski) [reply dali/imgcodec/decoders/decoder_test_helper.h]: Done

**Review by banasraf — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

File: dali/util/numpy.cc [hunk 4]
```diff
@@ -214,5 +214,27 @@ size_t HeaderData::nbytes() const {
...
 
+Tensor<CPUBackend> ReadTensor(InputStream *src) {
+  numpy::HeaderData header;
+  numpy::ParseHeader(header, src);
+  src->SeekRead(header.data_offset, SEEK_SET);
+
+  Tensor<CPUBackend> data;
+  data.Resize(header.shape, header.type());
+  Index ret = src->Read(static_cast<uint8_t*>(data.raw_mutable_data()), header.nbytes());
+  DALI_ENFORCE(ret == (Index)header.nbytes(), "Failed to read numpy file");
```
Comment (jantonguirao) [inline dali/util/numpy.cc:225]:
```suggestion
  auto ret = src->Read(static_cast<uint8_t*>(data.raw_mutable_data()), header.nbytes());
  DALI_ENFORCE(ret == header.nbytes(), "Failed to read numpy file");
```
those are both `size_t`. No need to cast?

Comment (szkarpinski) [reply dali/util/numpy.cc]: Fixed, thank you

**Review by jantonguirao — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**